esphome/components/esp32/preferences.cpp

```diff
@@ -97,12 +99,9 @@ class ESP32Preferences : public ESPPreferences {
   ESPPreferenceObject make_preference(size_t length, uint32_t type) override {
     auto *pref = new ESP32PreferenceBackend();  // NOLINT(cppcoreguidelines-owning-memory)
     pref->nvs_handle = nvs_handle;
-    current_offset += length;
 
-    uint32_t keyval = current_offset ^ type;
-    char keybuf[16];
-    snprintf(keybuf, sizeof(keybuf), "%d", keyval);
-    pref->key = keybuf;  // copied to std::string
+    uint32_t keyval = type;
+    pref->key = str_sprintf("%u", keyval);
```

Comment:
```suggestion
    pref->key = str_sprintf("%d", keyval);
```
Wouldn't this limit the impact. Everything at offset 0 should still get the same key at offset 0?
The fact that the key can be negative is a 'cosmetic' point for me and actually doesn't make a difference

Reply:
Sure. The other way around: would that remove the braking-change flag? If more than one value stored in flash the user won't be able to retrieve it anyway. So if the braking-change stays I think this cosmetic change would worth it to not get negative keys. But if you insist I can revert it.

Reply:
@jesserockz your call

Reply:
The breaking change will stay regardless because as long as there are multiple components saving, any after the first wont be restored after upgrading to the version with this code.

Reply:
Is there anything more to be done? OttoWinter was marked accidentally for review by changing branch to beta (I was trying to remove tuya flag) and back. So I think we should not wait to him.

Reply:
I'm confused, why is this a breaking change? Yeah the key function is changed, but before any tiny change to the YAML config (and new components being added to ESPHome codebase, an unrelated component's data size changing etc) would also have changed the key. So the key _changing_ per release is the default, no?

Reply:
Not for all components. Some like `total_daily_energy` and `climate` use the `object_id_hash` as the key so it can survive firmware updates.

Reply:
Idea: We could test in make_preference with nvs_get_blob if a value exists with and old key calculation algorithm. If so we will rekey it and save it with the new one. This shouldn't wear the flash as it will be done only once. Reading flash doesn't wear it. It definitely slows down make_preference but might save some headaches. 
After a few releases we can drop the checking part of the algorithm and give some update path to users to avoid flash data loss.
This needs some coding and testing though, making my day. Does this worth it?

Reply:
> Not for all components. Some like `total_daily_energy` and `climate` use the `object_id_hash` as the key so it can survive firmware updates.

Unless I'm mistaken, they use the `object_id_hash` as the *type*. But the NVS key is/was still generated from the offset variable.

So before, the entry was still invalidated by these two cases:
1. A new esphome version adds a new internal NVS setting somewhere (and thus shifts around all entries that are generated afterwards)
2. The user adds a new sensor before this ilone in their YAML (again, the keys afterwards are sgifted back and get new nvs keys)

Thus the NVS key was not _quite_ stable before either. I guess though in effect 1 doesn't happen often enough that a user expects the NVS settings to get lost.

Reply:
Ah yes, did not think about those cases.